<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.datasets import make_spd_matrix
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score


pd.options.display.float_format = '{:,.4f}'.format

df = pd.read_csv('/datasets/insurance.csv')

In [2]:
display(df)
display(df.info())
display(df.describe())

Пол  Возраст    Зарплата  Члены семьи  Страховые выплаты
0       1  41.0000 49,600.0000            1                  0
1       0  46.0000 38,000.0000            1                  1
2       0  29.0000 21,000.0000            0                  0
3       0  21.0000 41,700.0000            2                  0
4       1  28.0000 26,100.0000            0                  0
...   ...      ...         ...          ...                ...
4995    0  28.0000 35,700.0000            2                  0
4996    0  34.0000 52,400.0000            1                  0
4997    0  20.0000 33,900.0000            2                  0
4998    1  22.0000 32,700.0000            3                  0
4999    1  28.0000 40,600.0000            1                  0

[5000 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


None

Пол    Возраст    Зарплата  Члены семьи  Страховые выплаты
count 5,000.0000 5,000.0000  5,000.0000   5,000.0000         5,000.0000
mean      0.4990    30.9528 39,916.3600       1.1942             0.1480
std       0.5000     8.4408  9,900.0836       1.0914             0.4632
min       0.0000    18.0000  5,300.0000       0.0000             0.0000
25%       0.0000    24.0000 33,300.0000       0.0000             0.0000
50%       0.0000    30.0000 40,200.0000       1.0000             0.0000
75%       1.0000    37.0000 46,600.0000       2.0000             0.0000
max       1.0000    65.0000 79,000.0000       6.0000             5.0000

In [3]:
df.isna().sum()

Пол                  0
Возраст              0
Зарплата             0
Члены семьи          0
Страховые выплаты    0
dtype: int64

Данные не требуют дополнительной предобработки

## Умножение матриц

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Вопроc:**

Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии? (Её можно обучить заново.)

**Ответ:** При умножении матрицы признаков на обратимую матрицу, качество линейной регресии не изменится.

**Обоснование:**
    


$P$- обратимая, по определению, матрица преобразования.
 
$(AB)^T=B^TA^T$

$
(AB)^{-1}=B^{-1}A^{-1}
$

$
AA^{-1}=A^{-1}A=E
$ - руководствуемся этим тождеством при преобразовании выражения a'.

Для ответа необходимо доказать (или опровергнуть), что предсказания $a$ на исходных признаках будут равны $a_2$ вычесленному с учетом умножения признаков на заведомо обратимую матрицу (судя по исходным данным, это будет $P$).

1. Вычислим $w_2$, по результату применения преобразования P к  Подставим в выражение: 
 
 $w_2 = ((XP)^T XP)^{-1} (XP)^T y$
 
 
2. Раскроем скобки для транспонирования по свойству $(AB)^T = B^T A^T$
 
 $w_2 = (P^T X^T XP)^{-1} P^T X^T y$
 
3.Вынесем P за скобки по свойству $(AB)^{-1} = B^{-1} A^{-1}$

 $w_2 = P^{-1} (X^T X)^{-1} (P^T)^{-1} P^T X^T y$
 
4. Т.к. умножение матрицы на обратимую равно E, можем сократить часть $(P^T)^{-1} P^T$ до $E$ и сразу умножить на $X^T$ 
  
  $w_2 = P^{-1} (X^T X)^{-1} X^T y$
  
5. учитывая, что $w = (X^T X)^{-1} X^T y$ и произведем замену
 
 $w_2 = P^{-1} w$
 
**Теперь проверим признаки**

1. Запишем формулу
 
 $a_2 = XPw_2$
 
2. Подставим получившиеся значения
 
 $a_2 = XP P^{-1} w$
 
3. Опять же, вспомним, что умножение на обратную матрицу даст $E$
 
 $a_2 = Xw$
 
4. Учитывая, что $a = Xw$, получаем:
 
 $a_2 = a$


 $a = Xw = a' = XPw'$
    
 $((XP)^T XP)^{-1} (XP)^T y = (P^T X^T XP)^{-1} P^T X^T y = P^{-1}(X^T X)^{-1}(P^T)^{-1} P^T X^T y =
P^{-1}(X^TX)^{-1} X^T y = P^{-1}w$  

Таким образом веса нашей новой модели также будут умножены на обратную матрицу P, и мы получим такие же результаты как и раньше( мы видим что важным условием является обратимость матрицы P)

## Алгоритм преобразования

**Алгоритм**

1. Создадим случайную, обратимую матрицу $P$
2. Умножим признаки матрицы X на матрицу $P$. Получим преобразованную матрицу признаков $XP$.



**Обоснование**

 $XPP^{-1} = X$

Таким образом, с помощью обратимой матрицы преобразования $P$ мы можем как кодировать, так и декодировать признаки

## Проверка алгоритма

Отделим целевой признак

In [4]:
features = df.drop('Страховые выплаты', axis=1)
target = df['Страховые выплаты']

Создадим случайную квадратную матрицу преобразования , и проверим ее на обратимость

In [5]:
p_matrix = make_spd_matrix(n_dim=4, random_state=12345)
try:
    np.linalg.inv(p_matrix)
    print('Матрица обратима')
except:
    print('Матрица не обратима')

Матрица обратима


Получим новую матрицу признаков путем преобразования исходной матрицы признаков умножением на матрицу **p_matrix**.

In [6]:
features_p = features @ p_matrix

Проверим полученную новую матрицу признаков на возможность восстановления с помощью обратной матрицы преоразования.


In [7]:
display(features)
display(features_p @ np.linalg.inv(p_matrix))

Пол  Возраст    Зарплата  Члены семьи
0       1  41.0000 49,600.0000            1
1       0  46.0000 38,000.0000            1
2       0  29.0000 21,000.0000            0
3       0  21.0000 41,700.0000            2
4       1  28.0000 26,100.0000            0
...   ...      ...         ...          ...
4995    0  28.0000 35,700.0000            2
4996    0  34.0000 52,400.0000            1
4997    0  20.0000 33,900.0000            2
4998    1  22.0000 32,700.0000            3
4999    1  28.0000 40,600.0000            1

[5000 rows x 4 columns]

0       1           2      3
0     1.0000 41.0000 49,600.0000 1.0000
1     0.0000 46.0000 38,000.0000 1.0000
2    -0.0000 29.0000 21,000.0000 0.0000
3     0.0000 21.0000 41,700.0000 2.0000
4     1.0000 28.0000 26,100.0000 0.0000
...      ...     ...         ...    ...
4995 -0.0000 28.0000 35,700.0000 2.0000
4996 -0.0000 34.0000 52,400.0000 1.0000
4997 -0.0000 20.0000 33,900.0000 2.0000
4998  1.0000 22.0000 32,700.0000 3.0000
4999  1.0000 28.0000 40,600.0000 1.0000

[5000 rows x 4 columns]

Исходная матрица признаков идентично восстановлена с помощью обратной матрицы преобразования.

Разобьем на обучающие и тестовые выборки исходный и преобразованный датафреймы.

In [8]:
features_train, features_test, target_train, target_test = (
         train_test_split(features, target,test_size=0.25, random_state=12345))
features_p_train, features_p_test, target_p_train, target_p_test = (
         train_test_split(features, target,test_size=0.25, random_state=12345))

Обучим исходную и преобразованную модели на оригинальных и преобразованных данных.

In [9]:
model = LinearRegression()
model.fit(features_train, target_train)
prediction = model.predict(features_test)
print('R2 scor : {:.10f}'.format(r2_score(target_test, prediction),'\n'))
print('Веса:  пол  {:.4f}, возраст {:.4f}, зарплата {:.4f}, члены семьи {:.4f}'.format(*model.coef_))

R2 scor : 0.4352275713
Веса:  пол  0.0179, возраст 0.0357, зарплата -0.0000, члены семьи -0.0126


In [10]:
model = LinearRegression()
model.fit(features_p_train, target_p_train)
prediction_p = model.predict(features_p_test)
print('R2 scor : {:.10f}'.format(r2_score(target_p_test, prediction_p),'\n'))
print('Веса:  пол  {:.4f}, возраст {:.4f}, зарплата {:.4f}, члены семьи {:.4f}'.format(*model.coef_))

R2 scor : 0.4352275713
Веса:  пол  0.0179, возраст 0.0357, зарплата -0.0000, члены семьи -0.0126


Значения метрик не изменились.

# *Вывод*


1. Были изучены данные.
2. Доказано, что при умножении на обратимую матрицу качество линейной регрессии не изменится. 
3. Был запрограммирован алгоритм преобразования и проверено качество линейной регрессии после преобразования.
4. Метрика качества линейной регрессии R2 на исходных признаках: 00.4352275713.
5. Метрика качества линейной регрессии R2 на преобразованных признаках: 0.4352275713.
6. Таким образом, показано, что качество линейной регрессии не изменилось, метрика R2 не меняет совего значения при выбранной точности округления 10 знаков после запятой.
7. Можно сделать вывод,что алгоритм преобразования данных работает правильно.

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования